### ISSUE TITLE
Qt5: qtdeclarative pkg-config missing an include path

### ISSUE THREAD

**Issue by lrworth** [NixOS/nixpkgs#52457]
## Issue description

The pkg-config Cflags for qtdeclarative is missing `-I${includedir}`. It only has `-I${includedir}/QtQuick`. It appears that conventionally both of these paths appear in the include path, so that people write code like:
```
#include <QQmlParserStatus>
#include <QtQml/QQmlParserStatus>
```
where they can choose to put the "package" name in the include path.

qtbase already allows for this. The same problem may apply to all the other qt packages.

### Steps to reproduce

Note the Cflags in both files below:
```
➜  cat /nix/store/g6mfmwspn1n5gcryar0b8401fq6livsc-qtdeclarative-5.11.1-dev/lib/pkgconfig/Qt5Quick.pc
prefix=/nix/store/fsr4yhp9ks67bdrl24g34bh7vhvxv6a4-qtdeclarative-5.11.1
exec_prefix=/nix/store/lvxw5wal0flf1s1iqmrrnw33h6g7llky-qtdeclarative-5.11.1-bin
libdir=${prefix}/lib
includedir=/nix/store/g6mfmwspn1n5gcryar0b8401fq6livsc-qtdeclarative-5.11.1-dev/include


Name: Qt5 Quick
Description: Qt Quick module
Version: 5.11.1
Libs: -lQt5Quick 
Cflags: -DQT_QUICK_LIB -I${includedir}/QtQuick 
Requires: Qt5Core Qt5Gui Qt5Qml

➜  cat /nix/store/x2bdpznvin39baq9wif47ajcgf8in05w-qtbase-5.11.1-dev/lib/pkgconfig/Qt5Core.pc
prefix=/nix/store/azldmypwnamkxnm8hdggsf515i8x8b47-qtbase-5.11.1
exec_prefix=${prefix}
libdir=${prefix}/lib
includedir=/nix/store/x2bdpznvin39baq9wif47ajcgf8in05w-qtbase-5.11.1-dev/include

host_bins=/nix/store/x2bdpznvin39baq9wif47ajcgf8in05w-qtbase-5.11.1-dev/bin
qt_config=shared release c++11 c++14 c++1z concurrent dbus reduce_exports release_tools stl properties animation textcodec big_codecs codecs textdate datestring doubleconversion filesystemiterator filesystemwatcher gestures itemmodel proxymodel identityproxymodel library mimetype process statemachine regularexpression settings sharedmemory sortfilterproxymodel stringlistmodel systemsemaphore temporaryfile translation xmlstream xmlstreamreader xmlstreamwriter

Name: Qt5 Core
Description: Qt Core module
Version: 5.11.1
Libs: -L${libdir} -lQt5Core 
Cflags: -DQT_CORE_LIB -I${includedir}/QtCore -I${includedir}

```

## Technical details

 - system: `"x86_64-darwin"`
 - host os: `Darwin 18.2.0, macOS 10.14.1`
 - multi-user?: `no`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.1.3`
 - channels(lukeworth): `"nixpkgs-19.03pre161900.61c3169a0e1"`
 - nixpkgs: `/Users/lukeworth/.nix-defexpr/channels/nixpkgs`

**Comment by xbreak** [comment]
I ran into this problem as well while moc-ing.
Note that the Qt Headers themselves mainly use the include form `<QTMODULE/QTHEADER>` so Qt itself expects the include path `$prefix/include`.

Is this an upstream issue?

**Comment by jtojnar** [comment]
Yeah, this looks like boog – [Arch package](https://www.archlinux.org/packages/extra/x86_64/qt5-declarative/) has `-I${includedir}` in the .pc files and it does not look like they do anything special.

**Comment by jtojnar** [comment]
Looking at https://archive.archlinux.org/packages/q/qt5-declarative/, the key is present even in 5.12.4 so we must be breaking it somewhere.

**Comment by xbreak** [comment]
[CentOS 8](https://centos.pkgs.org/8/centos-appstream-x86_64/qt5-qtbase-devel-5.11.1-7.el8.i686.rpm.html) also adds `-I{includedir}` (using Qt-base as an example here).

**Comment by jtojnar** [comment]
Likely caused by this line: https://github.com/qt/qtbase/blob/b4e17a48646a1d2156b7d56ab6003db46af79cec/qmake/generators/makefile.cpp#L3391-L3392

**Comment by xbreak** [comment]
@jtojnar: Do you think we can build Qt differently so that the .pc-files are generated correctly?
It's not clear to me how the files are created.

**Comment by jtojnar** [comment]
So when the ¿nixpkgs’s cc-wrapper passes `-isystem ${dev}/include` to the compiler? (I thought it only does this for dependencies, maybe it comes from other place), Qt marks it as a system include path:

https://github.com/qt/qtbase/blob/3b4211ec47c98c09e2b96c77df1f8f0833c92793/mkspecs/features/toolchain.prf#L77-L82

and qmake’s pkg-config generator considers it unnecessary to add to the `.pc` file since it will be available anyway:

https://github.com/qt/qtbase/blob/b4e17a48646a1d2156b7d56ab6003db46af79cec/qmake/generators/makefile.cpp#L3391-L3392

And it is in fact right, the cc-wrapper will add `-isystem ${dev}/include`

https://github.com/NixOS/nixpkgs/blob/f7ac956a2b5b8c7f3a50c63547af493c68372169/pkgs/build-support/cc-wrapper/setup-hook.sh#L71-L73

so C/C++ compilers should have no problem finding the headers.

**Comment by jtojnar** [comment]
Confirmed using the patch below that cc-wrapper only includes the dependency include paths, not qtsvg’s own include path so I am still not sure how it gets into `QMAKE_DEFAULT_INCDIRS`.

```patch
--- a/pkgs/development/libraries/qt-5/qtModule.nix
+++ b/pkgs/development/libraries/qt-5/qtModule.nix
@@ -33,6 +33,7 @@ mkDerivation (args // {
 
     fixQtBuiltinPaths . '*.pr?'
   '';
+  preBuild = ''echo $NIX_CFLAGS_COMPILE; die'';
 
   postFixup = ''
     if [ -d "''${!outputDev}/lib/pkgconfig" ]; then
```

**Comment by jtojnar** [comment]
> @jtojnar: Do you think we can build Qt differently so that the .pc-files are generated correctly?
> It's not clear to me how the files are created.

There are two things I do not understand:

1. How does the `${qtsvg}/dev` get into the `QMAKE_DEFAULT_INCDIRS` variable when building `qtsvg`? (Assuming, I diagnosed the correct [qmake line](https://github.com/qt/qtbase/blob/b4e17a48646a1d2156b7d56ab6003db46af79cec/qmake/generators/makefile.cpp#L3391-L3392), maybe try patching the line to run unconditionally to verify.)
2. What causes a problem for you. The cc-wrapper make sure the `includedir` is visible to the compiler, even when `-I ${includedir}` is omitted from the `.pc` file (or there is no `.pc` file at all). Does MOC use something else than the compiler from `$CXX`? What is the error message?

I am not familiar with Qt at all so for further debugging, someone like @ttuegel will be needed.

**Comment by xbreak** [comment]
> What causes a problem for you. 

The moc-compiler has its own parser et.c. so is completely independent from $CXX AFAIK. What we do is to pass all include flags explicitly, as found using pkg-config. Which works on CentOS but not in Nix.
Actually, this applies in general for us as our internal packages are quite big, so we want to cherry-pick from buildInputs using things like pkg-config only the necessary bits for each library. I haven't managed (yet) to provide an unwrapped gcc with a standard environment, but that's the goal.

**Comment by xbreak** [comment]
Actually, this seems to work (at least in terms of adding the missing include path, I haven't tested it more):

```diff
diff --git a/pkgs/development/libraries/qt-5/modules/qttools.nix b/pkgs/development/libraries/qt-5/modules/qttools.nix
index 573472a8b5a..9a30ca3c845 100644
--- a/pkgs/development/libraries/qt-5/modules/qttools.nix
+++ b/pkgs/development/libraries/qt-5/modules/qttools.nix
@@ -13,6 +13,7 @@ qtModule {
         -e '/^cmake_linguist_config_version_file.input =/ s|$$\[QT_HOST_DATA.*\]|${getDev qtbase}|'
   '';

+  qmakeFlags = [ "QMAKE_DEFAULT_INCDIRS=" ];
   devTools = [
     "bin/qcollectiongenerator"
     "bin/linguist"

```

**Comment by ttuegel** [comment]
As @jtojnar [discovered](https://github.com/NixOS/nixpkgs/issues/52457#issuecomment-621231497), this is caused by the lack of a unified system include path in Nixpkgs. We could try to patch that statement so that `-I${includedir}` is added unconditionally; I doubt it would hurt anything.

**Comment by xbreak** [comment]
@ttuegel: I'll update #86317 to patch [makefile.cpp](https://github.com/qt/qtbase/blob/b4e17a48646a1d2156b7d56ab6003db46af79cec/qmake/generators/makefile.cpp#L3391-L3392) instead, ok?

**Comment by ttuegel** [comment]
@xbreak Sounds great, thanks!

**Comment by xbreak** [comment]
@lrworth: Did you ever try the fix from #86317?

**Comment by lrworth** [comment]
@xbreak Thanks for following up on this :) Unfortunately (fortunately?) I no longer work on Qt applications so I haven't needed to address this.

**Comment by xbreak** [comment]
@lrworth, this was likely fixed in #86317, unless you disagree I think you can close this ticket.

**Comment by lrworth** [comment]
Thanks 😃
